{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "interstate-calcium",
   "metadata": {},
   "source": [
    "<header style=\"padding:10px;background:#f9f9f9;border-top:3px solid #00b2b1\"><img id=\"Teradata-logo\" src=\"https://www.teradata.com/Teradata/Images/Rebrand/Teradata_logo-two_color.png\" alt=\"Teradata\" width=\"220\" align=\"right\" />\n",
    "\n",
    "<p style = 'font-size:28px;font-family:Arial;color:#E37C4D'><b>VantageCloud Lake Edition Base Data Loading</b></p>\n",
    "<hr>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "received-source",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the minimal number of packages\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import json\n",
    "import pandas as pd\n",
    "import teradatasql\n",
    "\n",
    "from teradataml import *\n",
    "\n",
    "access_key = ''\n",
    "secret = ''"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "automated-catalog",
   "metadata": {},
   "source": [
    "# Section 1 - Load Data #\n",
    "\n",
    "### For this demo, load two BFS Tables and one OFS table ###\n",
    "1. **Demo.Customers_BFS** Customer Dimension table for banking data set.\n",
    "2. **Demo.Accounts_Mapping_BFS** Dimension table for mapping Customers to Accounts Fraud Detection data set.\n",
    "3. **Demo_OFS.Txn_History** Transaction history table for use with the OAF standalone demo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "competent-billy",
   "metadata": {},
   "outputs": [
    {
     "ename": "OperationalError",
     "evalue": "[Version 17.20.0.15] [Session 18672] [Teradata Database] [Error 5628] Column numaccounts not found in D or TD_SYSFNLIB.READ_NOS_CONTRACT.\n at gosqldriver/teradatasql.formatError ErrorUtil.go:89\n at gosqldriver/teradatasql.(*teradataConnection).formatDatabaseError ErrorUtil.go:217\n at gosqldriver/teradatasql.(*teradataConnection).makeChainedDatabaseError ErrorUtil.go:233\n at gosqldriver/teradatasql.(*teradataConnection).processErrorParcel TeradataConnection.go:814\n at gosqldriver/teradatasql.(*TeradataRows).processResponseBundle TeradataRows.go:2284\n at gosqldriver/teradatasql.(*TeradataRows).executeSQLRequest TeradataRows.go:851\n at gosqldriver/teradatasql.newTeradataRows TeradataRows.go:708\n at gosqldriver/teradatasql.(*teradataStatement).QueryContext TeradataStatement.go:122\n at gosqldriver/teradatasql.(*teradataConnection).QueryContext TeradataConnection.go:1318\n at database/sql.ctxDriverQuery ctxutil.go:48\n at database/sql.(*DB).queryDC.func1 sql.go:1757\n at database/sql.withLock sql.go:3439\n at database/sql.(*DB).queryDC sql.go:1752\n at database/sql.(*Conn).QueryContext sql.go:2011\n at main.goCreateRows goside.go:775\n at _cgoexp_d46cb3509e30_goCreateRows _cgo_gotypes.go:387\n at runtime.cgocallbackg1 cgocall.go:316\n at runtime.cgocallbackg cgocall.go:235\n at runtime.cgocallback asm_amd64.s:994\n at runtime.goexit asm_amd64.s:1594",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [9], line 19\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 19\u001b[0m     \u001b[43mcur\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mDROP TABLE demo.Customers_EBS\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m/opt/conda/lib/python3.8/site-packages/teradatasql/__init__.py:705\u001b[0m, in \u001b[0;36mTeradataCursor.execute\u001b[0;34m(self, sOperation, params, ignoreErrors)\u001b[0m\n\u001b[1;32m    704\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m params:\n\u001b[0;32m--> 705\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecutemany\u001b[49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43msOperation\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mignoreErrors\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    707\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mtype\u001b[39m (params [\u001b[38;5;241m0\u001b[39m]) \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;28mlist\u001b[39m, \u001b[38;5;28mtuple\u001b[39m]:\n\u001b[1;32m    708\u001b[0m     \u001b[38;5;66;03m# Excerpt from PEP 249 DBAPI documentation:\u001b[39;00m\n\u001b[1;32m    709\u001b[0m     \u001b[38;5;66;03m#  The parameters may also be specified as list of tuples to e.g. insert multiple rows in a single\u001b[39;00m\n\u001b[1;32m    710\u001b[0m     \u001b[38;5;66;03m#  operation, but this kind of usage is deprecated: .executemany() should be used instead.\u001b[39;00m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.8/site-packages/teradatasql/__init__.py:952\u001b[0m, in \u001b[0;36mTeradataCursor.executemany\u001b[0;34m(self, sOperation, seqOfParams, ignoreErrors)\u001b[0m\n\u001b[1;32m    950\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[0;32m--> 952\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m OperationalError (sErr)\n\u001b[1;32m    954\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconnection\u001b[38;5;241m.\u001b[39mbTimingLog:\n",
      "\u001b[0;31mOperationalError\u001b[0m: [Version 17.20.0.15] [Session 18672] [Teradata Database] [Error 3807] Object 'demo.Customers_EBS' does not exist.\n at gosqldriver/teradatasql.formatError ErrorUtil.go:89\n at gosqldriver/teradatasql.(*teradataConnection).formatDatabaseError ErrorUtil.go:217\n at gosqldriver/teradatasql.(*teradataConnection).makeChainedDatabaseError ErrorUtil.go:233\n at gosqldriver/teradatasql.(*teradataConnection).processErrorParcel TeradataConnection.go:814\n at gosqldriver/teradatasql.(*TeradataRows).processResponseBundle TeradataRows.go:2284\n at gosqldriver/teradatasql.(*TeradataRows).executeSQLRequest TeradataRows.go:851\n at gosqldriver/teradatasql.newTeradataRows TeradataRows.go:708\n at gosqldriver/teradatasql.(*teradataStatement).QueryContext TeradataStatement.go:122\n at gosqldriver/teradatasql.(*teradataConnection).QueryContext TeradataConnection.go:1318\n at database/sql.ctxDriverQuery ctxutil.go:48\n at database/sql.(*DB).queryDC.func1 sql.go:1757\n at database/sql.withLock sql.go:3439\n at database/sql.(*DB).queryDC sql.go:1752\n at database/sql.(*Conn).QueryContext sql.go:2011\n at main.goCreateRows goside.go:775\n at _cgoexp_d46cb3509e30_goCreateRows _cgo_gotypes.go:387\n at runtime.cgocallbackg1 cgocall.go:316\n at runtime.cgocallbackg cgocall.go:235\n at runtime.cgocallback asm_amd64.s:994\n at runtime.goexit asm_amd64.s:1594",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [9], line 55\u001b[0m\n\u001b[1;32m     24\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m\n\u001b[1;32m     27\u001b[0m qry \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'''\u001b[39m\n\u001b[1;32m     28\u001b[0m \u001b[38;5;124mCREATE TABLE demo.Customers_EBS,\u001b[39m\n\u001b[1;32m     29\u001b[0m \u001b[38;5;124mSTORAGE = TD_NDSSTORAGE\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     52\u001b[0m \u001b[38;5;124m    ) WITH DATA;\u001b[39m\n\u001b[1;32m     53\u001b[0m \u001b[38;5;124m    \u001b[39m\u001b[38;5;124m'''\u001b[39m\n\u001b[0;32m---> 55\u001b[0m \u001b[43mcur\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mqry\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.8/site-packages/teradatasql/__init__.py:705\u001b[0m, in \u001b[0;36mTeradataCursor.execute\u001b[0;34m(self, sOperation, params, ignoreErrors)\u001b[0m\n\u001b[1;32m    702\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m (\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparams unexpected type \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat (\u001b[38;5;28mtype\u001b[39m (params)))\n\u001b[1;32m    704\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m params:\n\u001b[0;32m--> 705\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecutemany\u001b[49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43msOperation\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mignoreErrors\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    707\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mtype\u001b[39m (params [\u001b[38;5;241m0\u001b[39m]) \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;28mlist\u001b[39m, \u001b[38;5;28mtuple\u001b[39m]:\n\u001b[1;32m    708\u001b[0m     \u001b[38;5;66;03m# Excerpt from PEP 249 DBAPI documentation:\u001b[39;00m\n\u001b[1;32m    709\u001b[0m     \u001b[38;5;66;03m#  The parameters may also be specified as list of tuples to e.g. insert multiple rows in a single\u001b[39;00m\n\u001b[1;32m    710\u001b[0m     \u001b[38;5;66;03m#  operation, but this kind of usage is deprecated: .executemany() should be used instead.\u001b[39;00m\n\u001b[1;32m    711\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mexecutemany (sOperation, params, ignoreErrors)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.8/site-packages/teradatasql/__init__.py:952\u001b[0m, in \u001b[0;36mTeradataCursor.executemany\u001b[0;34m(self, sOperation, seqOfParams, ignoreErrors)\u001b[0m\n\u001b[1;32m    949\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m bIgnore:\n\u001b[1;32m    950\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[0;32m--> 952\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m OperationalError (sErr)\n\u001b[1;32m    954\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconnection\u001b[38;5;241m.\u001b[39mbTimingLog:\n\u001b[1;32m    955\u001b[0m     timingLog (\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexecutemany call to goCreateRows took \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m ms\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat ((time\u001b[38;5;241m.\u001b[39mtime () \u001b[38;5;241m-\u001b[39m dStartTime) \u001b[38;5;241m*\u001b[39m \u001b[38;5;241m1000.0\u001b[39m))\n",
      "\u001b[0;31mOperationalError\u001b[0m: [Version 17.20.0.15] [Session 18672] [Teradata Database] [Error 5628] Column numaccounts not found in D or TD_SYSFNLIB.READ_NOS_CONTRACT.\n at gosqldriver/teradatasql.formatError ErrorUtil.go:89\n at gosqldriver/teradatasql.(*teradataConnection).formatDatabaseError ErrorUtil.go:217\n at gosqldriver/teradatasql.(*teradataConnection).makeChainedDatabaseError ErrorUtil.go:233\n at gosqldriver/teradatasql.(*teradataConnection).processErrorParcel TeradataConnection.go:814\n at gosqldriver/teradatasql.(*TeradataRows).processResponseBundle TeradataRows.go:2284\n at gosqldriver/teradatasql.(*TeradataRows).executeSQLRequest TeradataRows.go:851\n at gosqldriver/teradatasql.newTeradataRows TeradataRows.go:708\n at gosqldriver/teradatasql.(*teradataStatement).QueryContext TeradataStatement.go:122\n at gosqldriver/teradatasql.(*teradataConnection).QueryContext TeradataConnection.go:1318\n at database/sql.ctxDriverQuery ctxutil.go:48\n at database/sql.(*DB).queryDC.func1 sql.go:1757\n at database/sql.withLock sql.go:3439\n at database/sql.(*DB).queryDC sql.go:1752\n at database/sql.(*Conn).QueryContext sql.go:2011\n at main.goCreateRows goside.go:775\n at _cgoexp_d46cb3509e30_goCreateRows _cgo_gotypes.go:387\n at runtime.cgocallbackg1 cgocall.go:316\n at runtime.cgocallbackg cgocall.go:235\n at runtime.cgocallback asm_amd64.s:994\n at runtime.goexit asm_amd64.s:1594"
     ]
    }
   ],
   "source": [
    "# /* -------------------------------------------------------- */\n",
    "# -- Perform this as SYSDBA.\n",
    "# -- Load demo.Customer_BFS from S3\n",
    "# /* -------------------------------------------------------- */\n",
    "\n",
    "# load vars json\n",
    "with open('vars.json', 'r') as f:\n",
    "    session_vars = json.load(f)\n",
    "\n",
    "name = session_vars['hierarchy']['SYSDBA']['username']\n",
    "pwd = session_vars['hierarchy']['SYSDBA']['password']\n",
    "\n",
    "with teradatasql.connect(host = session_vars['environment']['host'], \n",
    "                     user = name, \n",
    "                     password = pwd) as con:\n",
    "    cur = con.cursor()\n",
    "    \n",
    "    try:\n",
    "        cur.execute('DROP TABLE demo.Customers_BFS')\n",
    "    except Exception as e:\n",
    "        if str(e.args).find(\"3807\") >= 1:\n",
    "            pass\n",
    "        else:\n",
    "            raise\n",
    "            \n",
    "    \n",
    "        qry = f'''\n",
    "        CREATE TABLE demo.Customers_BFS,\n",
    "        STORAGE = TD_NDSSTORAGE\n",
    "        AS (\n",
    "            SELECT\n",
    "             id,\n",
    "             customer_identifier,\n",
    "             firstname,\n",
    "             lastname,\n",
    "             email,\n",
    "             phone,\n",
    "             birthday,\n",
    "             streetaddress,\n",
    "             city,\n",
    "             state,\n",
    "             zipcode,\n",
    "             latitude,\n",
    "             longitude,\n",
    "             numaccounts\n",
    "            FROM\n",
    "            (\n",
    "                LOCATION = '/s3/td-demos-data.s3.amazonaws.com/DEMO/CUSTOMERS_EBS/manifest.json'\n",
    "                AUTHORIZATION = '{{\"Access_ID\":\"{access_key}\",\"Access_Key\":\"{secret}\"}}'\n",
    "                MANIFEST = 'TRUE'\n",
    "            ) as D\n",
    "            ) WITH DATA;\n",
    "            '''\n",
    "        \n",
    "        cur.execute(qry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "infectious-forth",
   "metadata": {},
   "outputs": [],
   "source": [
    "# /* -------------------------------------------------------- */\n",
    "# -- Perform this as SYSDBA.\n",
    "# -- Load demo.Accounts_Mapping_BFS from source file\n",
    "# /* -------------------------------------------------------- */\n",
    "\n",
    "# load vars json\n",
    "with open('vars.json', 'r') as f:\n",
    "    session_vars = json.load(f)\n",
    "\n",
    "name = session_vars['hierarchy']['SYSDBA']['username']\n",
    "pwd = session_vars['hierarchy']['SYSDBA']['password']\n",
    "\n",
    "with teradatasql.connect(host = session_vars['environment']['host'], \n",
    "                     user = name, \n",
    "                     password = pwd) as con:\n",
    "    cur = con.cursor()\n",
    "    \n",
    "    try:\n",
    "        cur.execute('DROP TABLE Demo.Accounts_Mapping_BFS')\n",
    "    except Exception as e:\n",
    "        # Table already exists\n",
    "        if str(e.args).find(\"3807\") >= 1:\n",
    "            pass\n",
    "        else:\n",
    "            raise\n",
    "            \n",
    "    \n",
    "    qry = f'''\n",
    "    CREATE MULTISET TABLE demo.Accounts_Mapping_BFS\n",
    "    AS (\n",
    "        SELECT nameOrig,\n",
    "         customer_identifier\n",
    "        FROM\n",
    "        (\n",
    "            LOCATION = '/s3/td-demos-data.s3.amazonaws.com/DEMO/ACCOUNTS_MAPPING_EBS/manifest.json'\n",
    "            AUTHORIZATION = '{{\"Access_ID\":\"{access_key}\",\"Access_Key\":\"{secret}\"}}'\n",
    "            MANIFEST = 'TRUE'\n",
    "        ) as D\n",
    "        ) WITH DATA;\n",
    "    '''\n",
    "    cur.execute(qry)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ecological-refund",
   "metadata": {},
   "outputs": [],
   "source": [
    "# /* -------------------------------------------------------- */\n",
    "# -- Perform this as SYSDBA.\n",
    "# -- Create the transaction history OFS table for the OAF demo\n",
    "# /* -------------------------------------------------------- */\n",
    "\n",
    "# load vars json\n",
    "with open('vars.json', 'r') as f:\n",
    "    session_vars = json.load(f)\n",
    "\n",
    "name = session_vars['hierarchy']['SYSDBA']['username']\n",
    "pwd = session_vars['hierarchy']['SYSDBA']['password']\n",
    "\n",
    "with teradatasql.connect(host = session_vars['environment']['host'], \n",
    "                     user = name, \n",
    "                     password = pwd) as con:\n",
    "    cur = con.cursor()\n",
    "    \n",
    "    try:\n",
    "        cur.execute('DROP TABLE demo_ofs.Txn_History')\n",
    "    except Exception as e:\n",
    "        # Table already exists\n",
    "        if str(e.args).find(\"3807\") >= 1:\n",
    "            pass\n",
    "        elif str(e.args).find(\"4880\") >= 1:\n",
    "            pass\n",
    "        else:\n",
    "            raise\n",
    "            \n",
    "    \n",
    "    qry = '''\n",
    "    CREATE TABLE demo_ofs.Txn_History,\n",
    "\n",
    "        /*Note the use of STORAGE here to specify OFS\n",
    "        This can also be set to a user and database default */\n",
    "\n",
    "        STORAGE=TD_OFSSTORAGE\n",
    "\n",
    "        /* We can use a SELECT statement to populate the table\n",
    "        using WITH DATA */\n",
    "        AS (\n",
    "\n",
    "            /*Provide json path information to identify the proper fields */\n",
    "            SELECT Payload.txn_id txn_id,\n",
    "                Payload.step step,\n",
    "                Payload.\"type\" \"txn_type\",\n",
    "                CAST(Payload.amount AS FLOAT) amount,\n",
    "                Payload.nameOrig nameOrig,\n",
    "                CAST(Payload.oldbalanceOrig AS FLOAT) oldbalanceOrig,\n",
    "                CAST(Payload.newbalanceOrig AS FLOAT) newbalanceOrig,\n",
    "                Payload.nameDest nameDest,\n",
    "                CAST(Payload.oldbalanceDest AS FLOAT) oldbalanceDest,\n",
    "                CAST(Payload.newbalanceDest AS FLOAT) newbalanceDest,\n",
    "                CAST(Payload.isFraud AS INTEGER) isFraud,\n",
    "                CAST(Payload.isFlaggedFraud AS INTEGER) isFlaggedFraud\n",
    "\n",
    "            FROM (\n",
    "                LOCATION = '/s3/s3.amazonaws.com/trial-datasets/FraudReduction/'\n",
    "                AUTHORIZATION = repositories.PubAuth\n",
    "    ) AS D\n",
    "    ) WITH DATA\n",
    "    '''\n",
    "    try:\n",
    "        cur.execute(qry)\n",
    "    except Exception as e:\n",
    "        # Table already exists\n",
    "        if str(e.args).find(\"3803\") >= 1:\n",
    "            pass "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "orange-electron",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
